web-common/src/features/sources/modal/AddDataForm.svelte

```diff
@@ -80,57 +98,103 @@
         );
     }
   }
+
+  async function handleOnUpdate<
+    T extends Record<string, unknown>,
+    M = any,
+    In extends Record<string, unknown> = T,
+  >(event: {
+    form: SuperValidated<T, M, In>;
+    formEl: HTMLFormElement;
+    cancel: () => void;
+    result: Extract<ActionResult, { type: "success" | "failure" }>;
+  }) {
+    if (!event.form.valid) return;
+    const values = event.form.data;
+
+    try {
+      await submitAddDataForm(queryClient, formType, connector, values);
+      onClose();
+    } catch (e) {
+      if (useDsn) {
+        dsnRpcError = e?.response?.data;
+      } else {
+        rpcError = e?.response?.data;
+      }
+    }
+  }
 </script>
 
 <div class="h-full w-full flex flex-col">
-  <form
-    class="pb-5 flex-grow overflow-y-auto"
-    id={formId}
-    use:enhance
-    on:submit|preventDefault={submit}
-  >
-    <div class="pb-2 text-slate-500">
-      Need help? Refer to our
-      <a
-        href="https://docs.rilldata.com/build/connect"
-        rel="noreferrer noopener"
-        target="_blank">docs</a
-      > for more information.
+  <div class="pb-2 text-slate-500">
+    Need help? Refer to our
+    <a
+      href="https://docs.rilldata.com/build/connect"
+      rel="noreferrer noopener"
+      target="_blank">docs</a
+    > for more information.
+  </div>
+
+  {#if hasDsnFormOption}
+    <div class="py-3">
+      <div class="text-sm font-medium mb-2">Connection method</div>
+      <ButtonGroup
+        selected={[useDsn ? "dsn" : "parameters"]}
+        on:subbutton-click={handleConnectionTypeChange}
+      >
+        <SubButton value="parameters" ariaLabel="Enter parameters">
+          <span class="px-2">Enter parameters</span>
+        </SubButton>
+        <SubButton value="dsn" ariaLabel="Use connection string">
+          <span class="px-2">Enter connection string</span>
+        </SubButton>
+      </ButtonGroup>
     </div>
-    {#if rpcError}
-      <SubmissionError
-        message={humanReadableErrorMessage(
-          connector.name,
-          rpcError.code,
-          rpcError.message,
-        )}
-      />
-    {/if}
-
-    {#each properties as property (property.key)}
-      {#if property.key !== undefined && !property.noPrompt}
+  {/if}
+
+  {#if !useDsn}
```

Comment:
Not sure if this is our standard, but it feels weird to have 2 significantly independent form in a single component. Would feel cleaner to me if they were isolated in separate components.

Reply:
I agree it's not great, but I was hesitant to split them apart because:
1. We already have the "AddDataModal" wrapping the "AddDataForm". Adding another layer of abstraction would make it harder to follow.
2. The two forms share much in common, like the submission function, structure, and field types.
3. A change that applies to one of these forms likely applies to the other. So this follows "code that changes together lives together".

That being said, splitting them up still might be net-positive. Given there's changes coming to this form flow, I'd propose we keep as is for now and maybe split them up as the modal dialog evolves further.